File: forms/framingham-risk-score-for-hard-coronary-heart-disease/front-end-patient-form-with-svelte/src/lib/engine/risk-grader.ts
import type { AssessmentData, FiredRule, RiskLevel } from './types';
import { allRules } from './risk-rules';
import { calculateFraminghamRisk, isLikelyDraft } from './utils';

/**
 * Pure function: evaluates all risk rules against assessment data.
 * Returns the risk level, 10-year risk percentage, and fired rules.
 * If age and sex are missing, returns "draft" status.
 */
export function calculateRisk(data: AssessmentData): {
	riskCategory: RiskLevel;
	tenYearRiskPercent: number;
	firedRules: FiredRule[];
} {
	// Check if draft (age and sex missing)
	if (isLikelyDraft(data)) {
		return { riskCategory: 'draft', tenYearRiskPercent: 0, firedRules: [] };
	}

	// Calculate 10-year risk using Framingham Cox regression
	const rawRisk = calculateFraminghamRisk(data);

	// Round to 1 decimal place
	const tenYearRiskPercent = Math.round(rawRisk * 10) / 10;

	// Fire rules
	const firedRules: FiredRule[] = [];
	for (const rule of allRules()) {
		if (rule.evaluate(data, tenYearRiskPercent)) {
			firedRules.push({
				id: rule.id,
				category: rule.category,
				description: rule.description,
				riskLevel: rule.riskLevel
			});
		}
	}

	// Determine risk category from 10-year risk
	let riskCategory: RiskLevel;
	if (tenYearRiskPercent < 10.0) {
		riskCategory = 'low';
	} else if (tenYearRiskPercent < 20.0) {
		riskCategory = 'intermediate';
	} else {
		riskCategory = 'high';
	}

	return { riskCategory, tenYearRiskPercent, firedRules };
}

File: forms/framingham-risk-score-for-hard-coronary-heart-disease/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/risk_grader.rs
use super::risk_rules::all_rules;
use super::types::{AssessmentData, FiredRule, RiskLevel};
use super::utils::{calculate_framingham_risk, is_likely_draft};

/// Pure function: evaluates all risk rules against assessment data.
/// Returns the risk level, 10-year risk percentage, and fired rules.
/// If age and sex are missing, returns "draft" status.
pub fn calculate_risk(data: &AssessmentData) -> (RiskLevel, f64, Vec<FiredRule>) {
    // Check if draft (age and sex missing)
    if is_likely_draft(data) {
        return ("draft".to_string(), 0.0, vec![]);
    }

    // Calculate 10-year risk using Framingham Cox regression
    let risk_percent = calculate_framingham_risk(data);

    // Round to 1 decimal place
    let risk_percent = (risk_percent * 10.0).round() / 10.0;

    // Fire rules
    let mut fired_rules = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data, risk_percent) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                risk_level: rule.risk_level.to_string(),
            });
        }
    }

    // Determine risk category from 10-year risk
    let level = if risk_percent < 10.0 {
        "low"
    } else if risk_percent < 20.0 {
        "intermediate"
    } else {
        "high"
    };

    (level.to_string(), risk_percent, fired_rules)
}
